code/controllers/subsystems/supply.dm

```diff
@@ -121,67 +127,52 @@ SUBSYSTEM_DEF(supply)
 					if(AM.type in CC.wanted_types)
 						CC.Complete(1)
 						qdel(AM)
-						continue
+						break
+
+				if(QDELETED(AM))	//Our atom was qdel'd/used in a contract-- move onto the next atom to free up the reference for GC
+					continue
 
 			if(istype(AM, /obj/structure/closet/crate/))
 				var/obj/structure/closet/crate/CR = AM
-				callHook("sell_crate", list(CR, subarea))
-				add_points_from_source(CR.points_per_crate, "crate")
+				crates[CR] = subarea	//Store the crate to be handled later, as we'll have object refrences we'll need for the contents inside it
 				var/find_slip = 1
-				for(var/atom in CR)
-					// Sell manifests
-					var/atom/A = atom
 
+				for(var/atom/A in CR)
 					if(GLOB.using_map.using_new_cargo)
 						for(var/datum/contract/cargo/CC in GLOB.using_map.contracts) //just in case someone shoved it in a crate
 							if(A.type in CC.wanted_types)
 								CC.Complete(1)
-								//qdel(AM)
-								continue
-
-						if(istype(A, /obj/item/stack/material))
-							var/obj/item/stack/material/P = A
-							var/material/material = P.get_material()
-							if(material.sale_price > 0)
-								var/materialmoney = P.get_amount() * material.sale_price
-								materialmoney *= GLOB.using_map.new_cargo_inflation
-								materialmoney *= 0.25 //test and balance
-								material_count[material.display_name] += materialmoney
-
-						else
+								qdel(A)
+								break
 
-							var/obj/O = A
-							var/addvalue
+						if(QDELETED(A))
+							continue
 
-							if(GLOB.using_map.trading_faction)
-								if(GLOB.using_map.trading_faction.reputation > 50)
-									var/newvalue = (GLOB.using_map.trading_faction.reputation - 50) / 100
-									addvalue = (find_item_value(O) * ((newvalue / 2) + 0.75))
-							else
-								addvalue = (find_item_value(O) * 0.75) //we get even less for selling in bulk
-
-							add_points_from_source(addvalue, "trade")
+					//Calculate sales for each material
+					if(istype(A, /obj/item/stack/material))
+						var/obj/item/stack/material/S = A
+						var/amount = S.get_amount()
+						if(S.material.sale_price > 0)
+							material_count[S.material.display_name] += find_item_value(S, amount)
+						if(S.reinf_material?.sale_price > 0)
+							material_count[S.reinf_material.display_name] += find_item_value(S, amount, TRUE)
```

Comment:
Good catch. I'll get on that

Reply:
Actually, after having a quick look, there might be an issue with using `name` as opposed to `display_name`. That is that currently the keys of `material_count[]` is passed to `add_points_from_source()` as the source. This could cause weird and duplicate entries on the cargo statistics screen and could even be completely missed in the supply nano module `print_summary()` proc which iterates through `point_source_descriptions[]` which in turn uses global material display names (not names).

I could of course work around this by indexing by the material name and using a nested list to contain the display_name to pass back to the `add_points_from_source()` proc but that seems rather excessive? How likely would using `display_name` lead to causing an issue? It's also the variable the current system is using